simul/monitor/measure.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return err
+	}
+	log.Lvl3("Connected to sink:", addr)
+	global.sink = addr
+	global.connection = conn
+	global.encoder = json.NewEncoder(conn)
+	return nil
+}
+
+// RecordSingleMeasure sends the pair name - value to the monitor directly.
+func RecordSingleMeasure(name string, value float64) {
+	sm := newSingleMeasure(name, value)
+	sm.Record()
+}
+
+func newSingleMeasure(name string, value float64) *singleMeasure {
+	return &singleMeasure{
+		Name:  name,
+		Value: value,
+	}
+}
+
+func (s *singleMeasure) Record() {
+	if err := send(s); err != nil {
+		log.Error("Error sending SingleMeasure", s.Name, " to monitor:", err)
+	}
+}
+
+// NewTimeMeasure return *TimeMeasure
+func NewTimeMeasure(name string) *TimeMeasure {
+	tm := &TimeMeasure{name: name}
+	tm.reset()
+	return tm
+}
+
+// Record sends the measurements to the monitor:
+//
+// - wall time: *name*_wall
+//
+// - system time: *name*_system
+//
+// - user time: *name*_user
+func (tm *TimeMeasure) Record() {
+	// Wall time measurement
+	tm.Wall = newSingleMeasure(tm.name+"_wall", float64(time.Since(tm.lastWallTime))/1.0e9)
+	// CPU time measurement
+	tm.CPU.Value, tm.User.Value = getDiffRTime(tm.CPU.Value, tm.User.Value)
+	// send data
+	tm.Wall.Record()
+	tm.CPU.Record()
+	tm.User.Record()
+	// reset timers
+	tm.reset()
+
+}
+
+// reset reset the time fields of this time measure
+func (tm *TimeMeasure) reset() {
+	cpuTimeSys, cpuTimeUser := getRTime()
+	tm.CPU = newSingleMeasure(tm.name+"_system", cpuTimeSys)
+	tm.User = newSingleMeasure(tm.name+"_user", cpuTimeUser)
+	tm.lastWallTime = time.Now()
+}
+
+// Counter is an interface that can be used to report multiple values that
+// keeps evolving. The keys in the returned map is the name of the value to
+// record.
+type Counter interface {
+	Values() map[string]float64
+}
+
+// CounterMeasure is a struct that takes a Counter and can send the
+// measurements to the monitor. Each time Record() is called, the measurements
+// are put back to 0 (while the Counter still sends increased bytes number).
+type CounterMeasure struct {
+	name    string
+	counter Counter
+	baseMap map[string]float64
+}
+
+// NewCounterMeasure returns an CounterMeasure fresh. The base value are set
+// to the current value of counter.Rx() and counter.Tx()
+func NewCounterMeasure(name string, counter Counter) *CounterMeasure {
+	return &CounterMeasure{
+		name:    name,
+		counter: counter,
+		baseMap: counter.Values(),
+	}
+}
+
+// Record send the actual number of bytes read and written (**name**_written &
+// **name**_read) and reset the counters.
+func (cm *CounterMeasure) Record() {
+	newMap := cm.counter.Values()
+	for k, v := range cm.baseMap {
+		newV, ok := newMap[k]
+		if !ok {
+			continue
+		}
+		diff := newV - v
+		measure := newSingleMeasure(cm.name+"_"+k, diff)
+		measure.Record()
+		cm.baseMap[k] = newV
+	}
+}
+
+// Send transmits the given struct over the network.
+func send(v interface{}) error {
+	global.Lock()
```

Comment:
Since this is IO blocking function shall we Lock on the entire scope? What about protecting only "global.connection" by RLock?

Reply:
This package uses a bad engineering practice (global variable) but is super easy to use: there's one connection to a Sink for the whole package. The idea is that this package should only be used in simulations where you deploy independent binaries, so they each have "their own monitor package", so each one has a connection. The way I see it is ff you don't want to lock on entire scope, then you'd have to have separate connections for each measurement or stg much more complex... ?

Reply:
Ohh thx, I was missing the context